File: Windows/MediaFlowMonitor/LogEngine/DaVinciLogWatcher.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

namespace MediaFlowMonitor.LogEngine;

public enum ResolveLogSignalKind
{
    PluginCrash, GpuMemoryFull, DroppedFrames,
    RenderCacheRegenerated, FusionSlowNode, CodecSoftwareFallback, DbConnectionLost
}

public readonly record struct ResolveLogSignal(ResolveLogSignalKind Kind, string? Detail = null, int? Value = null);

/// Ascultă pasiv logurile DaVinci Resolve via FileSystemWatcher (event-driven,
/// echivalentul Windows al FSEventStream — zero polling).
public sealed class DaVinciLogWatcher : IDisposable
{
    private readonly string _logDirectory;
    private FileSystemWatcher? _watcher;
    private string? _activeLogPath;
    private long _lastOffset;

    public event EventHandler<ResolveLogSignal>? SignalDetected;

    public DaVinciLogWatcher()
    {
        _logDirectory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "Blackmagic Design", "DaVinci Resolve", "Support", "logs");
    }

    public void Start()
    {
        if (!Directory.Exists(_logDirectory)) return;

        _activeLogPath = FindLatestLogFile();
        if (_activeLogPath != null)
            _lastOffset = new FileInfo(_activeLogPath).Length;

        _watcher = new FileSystemWatcher(_logDirectory, "*.log")
        {
            NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.Size,
            EnableRaisingEvents = true
        };
        _watcher.Changed += OnLogChanged;
    }

    public void Stop()
    {
        _watcher?.Dispose();
        _watcher = null;
    }

    public void Dispose() => Stop();

    private string? FindLatestLogFile()
    {
        var files = Directory.GetFiles(_logDirectory, "*.log");
        return files.Length == 0 ? null : files[Array.IndexOf(files, files.MaxBy(File.GetLastWriteTimeUtc))];
    }

    private void OnLogChanged(object sender, FileSystemEventArgs e)
    {
        if (_activeLogPath == null) _activeLogPath = e.FullPath;
        if (e.FullPath != _activeLogPath) return;

        try
        {
            using var stream = new FileStream(e.FullPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            stream.Seek(_lastOffset, SeekOrigin.Begin);
            using var reader = new StreamReader(stream);
            string? line;
            while ((line = reader.ReadLine()) != null)
            {
                var signal = Parse(line);
                if (signal is { } s) SignalDetected?.Invoke(this, s);
            }
            _lastOffset = stream.Position;
        }
        catch (IOException)
        {
            // fișierul poate fi blocat momentan de Resolve — reîncercăm la următorul eveniment
        }
    }

    internal static ResolveLogSignal? Parse(string line)
    {
        if (line.Contains("GPU Memory Full"))
            return new ResolveLogSignal(ResolveLogSignalKind.GpuMemoryFull);
        if (line.Contains("Render Cache regenerated"))
            return new ResolveLogSignal(ResolveLogSignalKind.RenderCacheRegenerated);
        if (line.Contains("Database connection lost"))
            return new ResolveLogSignal(ResolveLogSignalKind.DbConnectionLost);
        if (line.Contains("decode fallback to software"))
            return new ResolveLogSignal(ResolveLogSignalKind.CodecSoftwareFallback);

        var crash = Regex.Match(line, @"OFX plugin .* crashed: (\w+)");
        if (crash.Success)
            return new ResolveLogSignal(ResolveLogSignalKind.PluginCrash, Detail: crash.Groups[1].Value);

        var dropped = Regex.Match(line, @"dropped (\d+) frames");
        if (dropped.Success)
            return new ResolveLogSignal(ResolveLogSignalKind.DroppedFrames, Value: int.Parse(dropped.Groups[1].Value));

        var fusion = Regex.Match(line, @"Fusion: comp render time (\d+)ms");
        if (fusion.Success)
            return new ResolveLogSignal(ResolveLogSignalKind.FusionSlowNode, Value: int.Parse(fusion.Groups[1].Value));

        return null;
    }
}

File: macOS/MediaFlowMonitor/Sources/MediaFlowMonitor/LogEngine/DaVinciLogWatcher.swift
import Foundation
import Combine

/// Semnale detectate în log-urile DaVinci Resolve.
enum ResolveLogSignal: Equatable {
    case pluginCrash(name: String)
    case gpuMemoryFull
    case droppedFrames(count: Int)
    case renderCacheRegenerated
    case fusionSlowNode(ms: Int)
    case codecSoftwareFallback(codec: String)
    case dbConnectionLost
}

/// Ascultă pasiv (FSEvents) fișierul de log activ al DaVinci Resolve
/// și publică semnale parsate, fără polling agresiv.
final class DaVinciLogWatcher {

    private var stream: FSEventStreamRef?
    private var lastOffset: UInt64 = 0
    private let logURL: URL
    let signalPublisher = PassthroughSubject<ResolveLogSignal, Never>()

    init?() {
        let base = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Blackmagic Design/DaVinci Resolve/logs")
        guard let latest = DaVinciLogWatcher.latestLogFile(in: base) else { return nil }
        self.logURL = latest
    }

    static func latestLogFile(in dir: URL) -> URL? {
        guard let files = try? FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: [.contentModificationDateKey]
        ) else { return nil }
        return files.filter { $0.pathExtension == "log" }
            .max { a, b in
                let da = (try? a.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
                let db = (try? b.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
                return da < db
            }
    }

    func start() {
        lastOffset = (try? FileManager.default.attributesOfItem(atPath: logURL.path)[.size] as? UInt64) ?? 0

        var context = FSEventStreamContext(
            version: 0,
            info: Unmanaged.passUnretained(self).toOpaque(),
            retain: nil, release: nil, copyDescription: nil
        )

        let callback: FSEventStreamCallback = { _, clientCallBackInfo, _, _, _, _ in
            guard let info = clientCallBackInfo else { return }
            let watcher = Unmanaged<DaVinciLogWatcher>.fromOpaque(info).takeUnretainedValue()
            watcher.readNewLines()
        }

        stream = FSEventStreamCreate(
            nil, callback, &context,
            [logURL.deletingLastPathComponent().path] as CFArray,
            FSEventStreamEventId(kFSEventStreamEventIdSinceNow),
            2.5, // latency ~2-3s, conform Regulii de polling pasiv
            FSEventStreamCreateFlags(kFSEventStreamCreateFlagFileEvents)
        )
        guard let stream else { return }
        FSEventStreamSetDispatchQueue(stream, DispatchQueue(label: "mediaflow.logwatcher", qos: .utility))
        FSEventStreamStart(stream)
    }

    func stop() {
        guard let stream else { return }
        FSEventStreamStop(stream)
        FSEventStreamInvalidate(stream)
        FSEventStreamRelease(stream)
        self.stream = nil
    }

    /// Citește doar octeții noi (tail), niciodată tot fișierul.
    private func readNewLines() {
        guard let handle = try? FileHandle(forReadingFrom: logURL) else { return }
        defer { try? handle.close() }
        try? handle.seek(toOffset: lastOffset)
        let data = handle.readDataToEndOfFile()
        lastOffset += UInt64(data.count)
        guard let text = String(data: data, encoding: .utf8), !text.isEmpty else { return }

        text.split(separator: "\n").forEach { line in
            if let signal = Self.parse(line: String(line)) {
                signalPublisher.send(signal)
            }
        }
    }

    static func parse(line: String) -> ResolveLogSignal? {
        if line.contains("GPU Memory Full") { return .gpuMemoryFull }
        if line.contains("Render Cache regenerated") { return .renderCacheRegenerated }
        if line.contains("Database connection lost") { return .dbConnectionLost }
        if let range = line.range(of: #"OFX plugin .* crashed: (\w+)"#, options: .regularExpression) {
            return .pluginCrash(name: String(line[range]))
        }
        if let range = line.range(of: #"dropped (\d+) frames"#, options: .regularExpression),
           let num = Int(line[range].filter(\.isNumber)) {
            return .droppedFrames(count: num)
        }
        if let range = line.range(of: #"Fusion: comp render time (\d+)ms"#, options: .regularExpression),
           let num = Int(line[range].filter(\.isNumber)) {
            return .fusionSlowNode(ms: num)
        }
        if line.contains("decode fallback to software") { return .codecSoftwareFallback(codec: "unknown") }
        return nil
    }
}
